react/src/hooks/useFeatureFlag.ts

```diff
@@ -0,0 +1,26 @@
+import { useEffect, useState } from 'react'
+import { JsonType } from 'posthog-js'
+import { usePostHog } from './usePostHog'
+
+export function useFeatureFlag<T = JsonType>(flag: string): T | undefined {
+    const client = usePostHog()
+
+    const [flagValue, setFlagValue] = useState<T | undefined>()
+    // would be nice to have a default value above however it's not possible due
+    // to a hydration error when using nextjs
+
+    useEffect(() => {
+        if (!client) {
+            return
+        }
+        return client.onFeatureFlags(() => {
+            if (client.getFeatureFlagPayload(flag)) {
+                setFlagValue(client.getFeatureFlagPayload(flag) as unknown as T | undefined)
+            } else {
+                setFlagValue(client.getFeatureFlag(flag) as unknown as T | undefined)
+            }
+        })
+    }, [client, flag])
+
+    return flagValue
+}
```

Comment:
Hot take - I think we should _only_ support this code path.

`useFeatureFlagEnabled` is the same as writing
```
const isEnabled = !!useFeatureFlag("flag")
```

`useFeatureFlags` avoids the callback code that we use for Experiments so is likely to get someone into a state they didn't intend.

`useFeatureFlagPayload` is identical to `useFeatureFlag` but with the case that it will respond with undefined if the key isn't a payload... I think this is super confusing. If I was the dev implementing this I wouldn't understand the difference between those two functions and would likely implement the wrong one.

Reply:
I don't like this, let's chat sync, sent an invite for Wed!

Reply:
I don't have a strong stance on here (and we can iterate), though there's another thing.

I didn't run the code, but looking at [Github](https://github.com/PostHog/posthog-js/blob/master/src/posthog-featureflags.ts#L276), `client.onFeatureFlags` doesn't return anything. So I'm not sure what's this `return client.onFeatureFlags()` returning.

From useEffect we should return a function that removes the feature flag listener. Otherwise when this component unmounts, posthog will still be calling the callback, and through it a useState setter on an unmounted component. Unfortunately posthog-js doesn't seem to support both attaching and detaching feature flag callbacks 🤔

Reply:
Ah this was probably a copy-paste from `posthog-js-lite`

Reply:
> From useEffect we should return a function that removes the feature flag listener. Otherwise when this component unmounts, posthog will still be calling the callback, and through it a useState setter on an unmounted component. Unfortunately posthog-js doesn't seem to support both attaching and detaching feature flag callbacks 🤔

I've added removal of the onFeatureFlags listener when the component is unmounted